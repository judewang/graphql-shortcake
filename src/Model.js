/* eslint no-param-reassign: ["error", { "ignorePropertyModificationsFor": ["target"] }] */
import _ from 'lodash';
import fp from 'lodash/fp';
import { invokeNull } from 'thelper';
import withArrayMutator from './withArrayMutator';
import withBuilder from './withBuilder';
import withBatch from './withBatch';
import withCache from './withCache';
import withColumn from './withColumn';
import withDefaultValues from './withDefaultValues';
import withFetcher from './withFetcher';
import withGlobalId from './withGlobalId';
import withIncrementer from './withIncrementer';
import withJSONMutator from './withJSONMutator';
import withLoader from './withLoader';
import withMutator from './withMutator';
import withSearcher from './withSearcher';

class Model {
  static database = null;

  static tableName = null;

  static viewName = null;

  static idAttribute = 'id';

  static hasOperator = false;

  static hasTimestamps = true;

  static softDelete = true;

  static get(target, name) {
    if (this.has(target, name)) {
      return target.get(name);
    }

    return target[name];
  }

  static set(target, name, value) {
    if (this.has(target, name)) {
      target.set(name, value);
      return;
    }

    target[name] = value;
  }

  static get displayName() {
    return this.name;
  }

  static has(target, name) {
    return target.ownKeys.indexOf(name) > -1;
  }

  static ownKeys(target) {
    return target.ownKeys;
  }

  static getOwnPropertyDescriptor(target, name) {
    if (this.has(target, name)) {
      return { configurable: true, enumerable: true, value: target.get(name) };
    }

    return Object.getOwnPropertyDescriptor(target, name);
  }

  static refreshView = _.debounce(() => {
    this.raw(`REFRESH MATERIALIZED VIEW ${this.viewName};`);
  }, 500, { maxWait: 2000 });

  static raw(...args) {
    return this.database.raw(...args);
  }

  static forge(attributes, options) {
    const model = new this({}, options);
    return model.forge(attributes);
  }

  static fromModel(model) {
    return _.get(model, ['nativeId'], model);
  }

  static checkOperator(operator) {
    const { hasOperator } = this;
    if (!hasOperator) return null;
    if (!operator) throw new Error('operator is required');
    return this.fromModel(operator);
  }

  [Symbol.toStringTag] = 'Model';

  [Symbol.toPrimitive](hint) {
    if (hint === 'number') return Number.NaN;
    return this.toString();
  }

  constructor() {
    if (this.initialize) this.initialize();
    const handler = _.pick(this.constructor, [
      'get', 'set', 'has', 'ownKeys', 'getOwnPropertyDescriptor',
    ]);
    return new Proxy(this, handler);
  }

  initialize() {
    this._ = {};
    this.current = {};
    this.previous = {};
  }

  get isNew() {
    const { idAttribute } = this.constructor;
    return !this[idAttribute];
  }

  get values() {
    return { ...this.previous, ...this.current };
  }

  get(paths) {
    const { values } = this;
    return _.get(values, paths, null);
  }

  set(paths, value) {
    _.set(this.current, paths, value);
  }

  valueOf(name) {
    if (name) return _.get(this.current, [name], this.previous[name]);

    const { idAttribute } = this.constructor;
    const id = this.valueOf(idAttribute);
    return id ? String(id) : id;
  }

  toString() {
    const { idAttribute } = this.constructor;
    return this[idAttribute];
  }

  clone(options) {
    const model = new this.constructor({}, options);
    model.current = _.cloneDeep(this.current);
    model.previous = this.previous;
    return model;
  }

  forge(attributes) {
    return invokeNull((attrs) => {
      const values = _.get(attrs, ['previous'], attrs);
      this.current = _.cloneDeep(values);
      this.previous = values;
      return this;
    }, attributes);
  }

  merge(attributes) {
    return invokeNull((attrs) => {
      const { format } = this.constructor;
      const values = format(attrs);
      _.forEach([this.current, this.previous], data => _.assign(data, values));
      return this;
    }, attributes);
  }
}

export default fp.compose(
  withArrayMutator,
  withJSONMutator,
  withIncrementer,
  withSearcher,
  withDefaultValues,
  withCache,
  withLoader,
  withFetcher,
  withMutator,
  withBatch,
  withBuilder,
  withColumn,
  withGlobalId,
)(Model);
