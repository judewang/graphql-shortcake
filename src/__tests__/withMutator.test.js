import { client } from 'knex';
import database from './database';
import Model from '..';

class Mutator extends Model {
  static database = database;

  static tableName = 'mutator';

  static hasOperator = true;

  static columns = {
    name: { type: String },
  }
}

class Logger extends Mutator {
  static hasOperator = false;

  static hasTimestamps = false;
}

class Cache extends Mutator {
  static softDelete = false;

  static hasOperator = false;

  static hasTimestamps = false;
}

describe('Mutator', () => {
  beforeAll(async () => {
    await database.schema.dropTableIfExists('mutator');

    await database.schema.createTable('mutator', (table) => {
      table.increments();
      table.string('name');
      table.timestamps();
      table.integer('created_by');
      table.integer('updated_by');
      table.timestamp('deleted_at');
      table.integer('deleted_by');
    });
  });

  describe('sent insert of save', () => {
    it('has operator', async () => {
      const model = new Mutator({ name: 'new is insert' });
      expect(model.values).toEqual({ name: 'new is insert' });
      await model.save(10);
      expect(model.values).toEqual(expect.objectContaining({
        createdAt: expect.anything(Date),
        createdBy: 10,
        updatedAt: expect.anything(Date),
        updatedBy: 10,
      }));
      expect(client).toMatchSnapshot();

      await expect(new Mutator().save())
        .rejects.toEqual(new Error('operator is required'));
    });

    it('no operator', async () => {
      const model = new Cache({ name: 'new is no operator insert' });
      await model.save();
      expect(model.values).toEqual(expect.objectContaining({
        createdAt: null,
        createdBy: null,
        updatedAt: null,
        updatedBy: null,
      }));
      expect(client).toMatchSnapshot();
    });
  });

  describe('sent update of save', async () => {
    it('has operator', async () => {
      const model = await Mutator.load(1);

      model.name = 'name is update';
      await model.save(20);
      await model.save(20);

      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(2);

      model.set({ name: 'update again' });
      await expect(model.save())
        .rejects.toEqual(new Error('operator is required'));
    });

    it('no operator', async () => {
      const model = await Cache.load(2);
      model.set({ name: 'name is no operator update' });
      await model.save();
      expect(client).toMatchSnapshot();
    });
  });

  describe('destroy', () => {
    it('has operator', async () => {
      const model = await Mutator.load(1);

      await model.destroy(10);
      await model.destroy(10);
      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(2);

      await expect(model.destroy())
        .rejects.toEqual(new Error('operator is required'));
    });

    it('no operator and no soft delete', async () => {
      const model = await Cache.load(2);
      await model.destroy();
      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(2);
    });

    it('no operator', async () => {
      const model = new Logger({ name: 'new logger' });
      await model.save();
      await model.destroy();
      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(2);
    });
  });
});
