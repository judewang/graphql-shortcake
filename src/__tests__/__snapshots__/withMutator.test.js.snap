// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutator destroy has operator 1`] = `
Array [
  "select * from mutator where id in (1) and deleted_at is null",
  "update mutator set deleted_at = DATE, deleted_by = 10 where deleted_at is null and id = 1 returning *",
]
`;

exports[`Mutator destroy no operator 1`] = `
Array [
  "insert into mutator (name) values (new logger) returning *",
  "update mutator set deleted_at = DATE where deleted_at is null and id = 3 returning *",
]
`;

exports[`Mutator destroy no operator and no soft delete 1`] = `
Array [
  "select * from mutator where id in (2)",
  "delete from mutator where id = 2",
]
`;

exports[`Mutator sent insert of save has operator 1`] = `
Array [
  "insert into mutator (created_at, created_by, name, updated_at, updated_by) values (DATE, 10, new is insert, DATE, 10) returning *",
]
`;

exports[`Mutator sent insert of save no operator 1`] = `
Array [
  "insert into mutator (name) values (new is no operator insert) returning *",
]
`;

exports[`Mutator sent update of save has operator 1`] = `
Array [
  "select * from mutator where id in (1) and deleted_at is null",
  "update mutator set name = name is update, updated_by = 20, updated_at = DATE where deleted_at is null and id = 1 returning *",
]
`;

exports[`Mutator sent update of save no operator 1`] = `
Array [
  "select * from mutator where id in (2)",
  "update mutator set name = name is no operator update where id = 2 returning *",
]
`;
